ghcide/src/Development/IDE/Core/Rules.hs

```diff
@@ -526,32 +514,56 @@ getHieAstsRule =
       hsc <- hscEnv <$> use_ GhcSession f
       getHieAstRuleDefinition f hsc tmr
 
+persistentHieFileRule :: Rules ()
+persistentHieFileRule = addPersistentRule GetHieAst $ \file -> runMaybeT $ do
+  res <- readHieFileForSrcFromDisk file
+  vfs <- asks vfs
+  (currentSource,ver) <- liftIO $ do
+    mvf <- getVirtualFile vfs $ filePathToUri' file
+    case mvf of
+      Nothing -> (,Nothing) <$> T.readFile (fromNormalizedFilePath file)
+      Just vf -> pure $ (Rope.toText $ _text vf, Just $ _lsp_version vf)
+  encoding <- liftIO $ getLocaleEncoding
+  let refmap = generateReferencesMap . getAsts . hie_asts $ res
+      del = deltaFromDiff (T.decode encoding $ hie_hs_src res) currentSource
+  pure $ (HAR (hie_module res) (hie_asts res) refmap mempty (HieFromDisk res),del,ver)
+
 getHieAstRuleDefinition :: NormalizedFilePath -> HscEnv -> TcModuleResult -> Action (IdeResult HieAstResult)
 getHieAstRuleDefinition f hsc tmr = do
   (diags, masts) <- liftIO $ generateHieAsts hsc tmr
+  se <- getShakeExtras
 
   isFoi <- use_ IsFileOfInterest f
   diagsWrite <- case isFoi of
     IsFOI Modified -> pure []
     _ | Just asts <- masts -> do
           source <- getSourceFileSource f
-          liftIO $ writeHieFile hsc (tmrModSummary tmr) (tcg_exports $ tmrTypechecked tmr) asts source
+          let exports = tcg_exports $ tmrTypechecked tmr
+              msum = tmrModSummary tmr
+          liftIO $ writeAndIndexHieFile hsc se msum f exports asts source
     _ -> pure []
```

Comment:
Don't we want to index FOIs too?

Reply:
We index them when we can be sure they match the state on disk (i.e. when they are saved).